app/com/linkedin/drelephant/spark/legacydata/SparkJobProgressData.java

```diff
@@ -0,0 +1,273 @@
+import org.apache.log4j.Logger;
+
+
+/**
+ * This class represents information contained in a job runtime process.
+ */
+public class SparkJobProgressData {
+  private static final Logger logger = Logger.getLogger(SparkJobProgressData.class);
+  private final Map<Integer, JobInfo> _jobIdToInfo = new HashMap<Integer, JobInfo>();
+  private final Set<Integer> _completedJobs = new HashSet<Integer>();
+  private final Set<Integer> _failedJobs = new HashSet<Integer>();
+
+  private final Map<StageAttemptId, StageInfo> _stageIdToInfo = new HashMap<StageAttemptId, StageInfo>();
+  private final Set<StageAttemptId> _completedStages = new HashSet<StageAttemptId>();
+  private final Set<StageAttemptId> _failedStages = new HashSet<StageAttemptId>();
+
+  public void addJobInfo(int jobId, JobInfo info) {
+    _jobIdToInfo.put(jobId, info);
+  }
+
+  public void addCompletedJob(int jobId) {
+    _completedJobs.add(jobId);
+  }
+
+  public void addFailedJob(int jobId) {
+    _failedJobs.add(jobId);
+  }
+
+  public void addStageInfo(int stageId, int attemptId, StageInfo info) {
+    _stageIdToInfo.put(new StageAttemptId(stageId, attemptId), info);
+  }
+
+  public void addCompletedStages(int stageId, int attemptId) {
+    _completedStages.add(new StageAttemptId(stageId, attemptId));
+  }
+
+  public void addFailedStages(int stageId, int attemptId) {
+    _failedStages.add(new StageAttemptId(stageId, attemptId));
+  }
+
+  public Set<Integer> getJobIds() {
+    return _jobIdToInfo.keySet();
+  }
+
+  public Set<StageAttemptId> getStageAttemptIds() {
+    return _stageIdToInfo.keySet();
+  }
+
+  public Set<Integer> getCompletedJobs() {
+    return _completedJobs;
+  }
+
+  public Set<Integer> getFailedJobs() {
+    return _failedJobs;
+  }
+
+  private static double getFailureRate(int numCompleted, int numFailed) {
+    int num = numCompleted + numFailed;
+
+    if (num == 0) {
+      return 0d;
+    }
+
+    return numFailed * 1.0d / num;
+  }
+
+  public double getJobFailureRate() {
+    return getFailureRate(_completedJobs.size(), _failedJobs.size());
+  }
+
+  public double getStageFailureRate() {
+    return getFailureRate(_completedStages.size(), _failedStages.size());
+  }
+
+  public JobInfo getJobInfo(int jobId) {
+    return _jobIdToInfo.get(jobId);
+  }
+
+  public StageInfo getStageInfo(int stageId, int attemptId) {
+    return _stageIdToInfo.get(new StageAttemptId(stageId, attemptId));
+  }
+
+  public Set<StageAttemptId> getCompletedStages() {
+    return _completedStages;
+  }
+
+  public Set<StageAttemptId> getFailedStages() {
+    return _failedStages;
+  }
+
+  /**
+   * Job itself does not have a name, it will use its latest stage as the name.
+   *
+   * @param jobId
+   * @return
+   */
+  public String getJobDescription(int jobId) {
+    List<Integer> stageIds = _jobIdToInfo.get(jobId).stageIds;
+    int id = -1;
+    for (int stageId : stageIds) {
+      id = Math.max(id, stageId);
+    }
+    if (id == -1) {
```

Comment:
This is just the empty-list case, right? Maybe write it explicitly instead of checking for init value.

Also, for nonempty lists, you could replace the loop with `Ordering` from Guava:

```
int id = Ordering.<Integer> natural().max(stageIds);
```